### ISSUE TITLE
Annotate.el likely has issues with doom

### ISSUE THREAD

**Issue by cage2** [hlissner/doom-emacs#4783]
Hi!

I am the co-maintainer of annotate.el a mode to annotate files without modify them.

https://github.com/bastibe/annotate.el/pulls

Two users noticed, some days ago, that this package does not works very well when Doom is used.

I can confirm that the bug appears only when using doom.

**What did you expect to happen?**

After loading said package a region should be annotated.

**What actually happened?**

No annotation appears and an error message is printed, instead:

```annotate-annotations-overlay-in-range: Symbol’s function definition is void: annotation```

**Additional details:**

- here is the original report:
  https://github.com/bastibe/annotate.el/issues/97

- the error printed is:

```annotate-annotations-overlay-in-range: Symbol’s function definition is void: annotation```

- backtrace

```lisp
Debugger entered--Lisp error: (void-function annotation)
  annotation()
  annotate-annotations-overlay-in-range(1 4)
  annotate-annotations-chain-in-range(1 4)
  annotate-annotate()
  funcall-interactively(annotate-annotate)
  call-interactively(annotate-annotate nil nil)
  command-execute(annotate-annotate)
```

**Steps to reproduce:**

- Load package annotate.el
- visit a non empty (and non org-mode) file and activate annotate-minor mode `M-x annotate-mode`
- mark a region of the text 
- type the command `C-c C-a`
- type the text of the annotation in the minibuffer
- an annotation should appears on the margin of the buffer or below the last annotated line  (depending of the length of the line that contains the region to be annotated)

**System information:**
https://pastebin.com/BWCrz0rS

Thanks!
C.

**Comment by hlissner** [comment]
This is a byte-compilation error. The `eannotate-ensure-annotation` macro isn't loaded while the package is byte-compiled, and is therefore treated like a function call (i.e. its arguments are evaluated).

The cause of this is a combination of:

1. **Expansion order of macros.** Macros are expanded at compile time in sequence order. This includes `cl-loop`, `cl-defmacro`, and `annotate-annotations-overlay-in-range`. The first time `annotate-annotations-overlay-in-range` is used [is before it is defined](https://github.com/bastibe/annotate.el/blob/master/annotate.el#L1165) (`cl-loop`).

   This results in the following byte-compiler warning:

   ```
    In annotate-ensure-annotation:
    annotate.el:1640:14: Warning: macro ‘annotate-ensure-annotation’ defined too
        late
   ```

   And can later result in the error that users have reported to you. The quick fix is to move the definition of `annotate-annotations-overlay-in-range` above the definition of `annotate-annotations-overlay-in-range`.

2. A vestigial advice in Doom was compiling packages with too little pre-initialized state. This was fixed in 2631932.

That said, even with no. 2, most users won't see this issue they're using the package manager built into Emacs (`package.el`), which loads the package before byte-compiling it, which causes the macro to be defined _before_ byte-compilation begins. This is not the case if the package is byte-compiled before it is loaded.

In any case, the issue should be resolved. Let me know if that isn't the case and I'll reopen it. Thanks for bringing it to my attention!

**Comment by cage2** [comment]
On Thu, Mar 11, 2021 at 10:06:22AM -0800, Henrik Lissner wrote:

Hi!

> This is a  byte-compilation error. The `eannotate-ensure-annotation`
> macro  isn't  loaded while  the  package  is byte-compiled,  and  is
> therefore  treated like  a  function call  (i.e.  its arguments  are
> evaluated).

And this totally makes sense to me.

Thank you very much, for your help! I suspected this issue was related
to byte compilation as  I just load and do not  compile the package on
my system. So you found the bug and provided the solution also! :)

Thanks a lot!
Bye!
C.
